### PR TITLE
feat: builder pattern for InstrumentedHttpClientConnectionManager

### PR DESCRIPTION
**Pull request by snakeru**
For my project I needed to specify only the name (and metrics registry, of course) and leave all the rest  at defaults. So here is the more flexible approach.

Also, the code was auto-formatted with IntelliJ.

### CODE REVIEW

File: metrics-httpclient/src/main/java/com/codahale/metrics/httpclient/InstrumentedHttpClientConnectionManager.java [hunk 1]
```diff
@@ -30,80 +30,48 @@
 
     protected static Registry<ConnectionSocketFactory> getDefaultRegistry() {
         return RegistryBuilder.<ConnectionSocketFactory>create()
-            .register("http", PlainConnectionSocketFactory.getSocketFactory())
-            .register("https", SSLConnectionSocketFactory.getSocketFactory())
-            .build();
+                .register("http", PlainConnectionSocketFactory.getSocketFactory())
```
Comment (arteam) [inline metrics-httpclient/src/main/java/com/codahale/metrics/httpclient/InstrumentedHttpClientConnectionManager.java]: Please don't add unrelated whitespace changes to your PR. This makes porting commits across different versions harder, because it increases the change of a possible merge conflict.

File: metrics-httpclient/src/main/java/com/codahale/metrics/httpclient/InstrumentedHttpClientConnectionManager.java [hunk 2]
```diff
@@ -30,80 +30,48 @@
...
+                .register("https", SSLConnectionSocketFactory.getSocketFactory())
+                .build();
     }
 
     private final MetricRegistry metricsRegistry;
     private final String name;
 
-    public InstrumentedHttpClientConnectionManager(MetricRegistry metricRegistry) {
```
Comment (arteam) [inline metrics-httpclient/src/main/java/com/codahale/metrics/httpclient/InstrumentedHttpClientConnectionManager.java]: We try to maintain backward compatibility with the previous version of Metrics. Please don't remove the public constructors, otherwise anybody who uses them will be forced to migrate their code to `Builder`, which should be optional.

File: metrics-httpclient/src/main/java/com/codahale/metrics/httpclient/InstrumentedHttpClientConnectionManager.java [hunk 3]
```diff
@@ -30,80 +30,48 @@
...
-    public InstrumentedHttpClientConnectionManager(MetricRegistry metricsRegistry,
+    private InstrumentedHttpClientConnectionManager(MetricRegistry metricsRegistry,
                                                    HttpClientConnectionOperator httpClientConnectionOperator,
                                                    HttpConnectionFactory<HttpRoute, ManagedHttpClientConnection>
                                                            connFactory,
                                                    long connTTL,
                                                    TimeUnit connTTLTimeUnit,
                                                    String name) {
         super(httpClientConnectionOperator, connFactory, connTTL, connTTLTimeUnit);
+        if (metricsRegistry==null) {
```
Comment (arteam) [inline metrics-httpclient/src/main/java/com/codahale/metrics/httpclient/InstrumentedHttpClientConnectionManager.java]: This is a good defensive pattern, but across the codebase we don't use it. It's expected that the user will provide a not-null `MetricsRegistry`. Otherwise, one can ask why do we check the metrics registry for nullability, but not `httpClientConnectionOperator` or `connTTLTimeUnit`? I think in the long run a better approach for the project would be to use `@Nullable` annotations from the checker framework for marking nullable fields and parameters, but that's out of the scope of this PR.

File: metrics-httpclient/src/main/java/com/codahale/metrics/httpclient/InstrumentedHttpClientConnectionManager.java [hunk 4]
```diff
@@ -30,80 +30,48 @@
...
+        }
         this.metricsRegistry = metricsRegistry;
         this.name = name;
 
         metricsRegistry.register(name(HttpClientConnectionManager.class, name, "available-connections"),
-            (Gauge<Integer>) () -> {
-                // this acquires a lock on the connection pool; remove if contention sucks
-                return getTotalStats().getAvailable();
-            });
+                (Gauge<Integer>) () -> {
```
Comment (arteam) [inline metrics-httpclient/src/main/java/com/codahale/metrics/httpclient/InstrumentedHttpClientConnectionManager.java]: The same comment about whitespacing changes. Also, please take a look at the warnings produced by `checkstyle`.

File: metrics-httpclient/src/main/java/com/codahale/metrics/httpclient/InstrumentedHttpClientConnectionManager.java [hunk 5]
```diff
@@ -114,4 +82,76 @@ public void shutdown() {
...
+        private SchemePortResolver schemePortResolver;
+        private DnsResolver dnsResolver = SystemDefaultDnsResolver.INSTANCE;
+        private long connTTL = -1;
+        private TimeUnit connTTLTimeUnit = TimeUnit.MILLISECONDS;
+        private String name;
+
+        Builder() {
+        }
+
+        public Builder metricsRegistry(final MetricRegistry metricsRegistry) {
```
Comment (arteam) [inline metrics-httpclient/src/main/java/com/codahale/metrics/httpclient/InstrumentedHttpClientConnectionManager.java]: To be honest, I don't see much value in declaring the method parameters as `final`. It's clear from the semantic of `Builder` that they are not going to change and we don't force final method parameters in the codestyle. I would remove them unless there's a specific reason to keep.

File: metrics-httpclient/src/main/java/com/codahale/metrics/httpclient/InstrumentedHttpClients.java [hunk 6]
```diff
@@ -28,8 +28,7 @@ public static HttpClientBuilder custom(MetricRegistry metricRegistry,
                                            HttpClientMetricNameStrategy metricNameStrategy) {
         return HttpClientBuilder.create()
                 .setRequestExecutor(new InstrumentedHttpRequestExecutor(metricRegistry, metricNameStrategy))
-                .setConnectionManager(new InstrumentedHttpClientConnectionManager(metricRegistry));
+                .setConnectionManager(InstrumentedHttpClientConnectionManager.builder(metricRegistry).build());
```
Comment (arteam) [inline metrics-httpclient/src/main/java/com/codahale/metrics/httpclient/InstrumentedHttpClients.java:31]: This change doesn't seem to be very useful, at the end we invoke the same constructor anyway...

File: metrics-httpclient/src/main/java/com/codahale/metrics/httpclient/InstrumentedHttpClientConnectionManager.java [hunk 7]
```diff
@@ -114,4 +82,76 @@ public void shutdown() {
         metricsRegistry.remove(name(HttpClientConnectionManager.class, name, "max-connections"));
         metricsRegistry.remove(name(HttpClientConnectionManager.class, name, "pending-connections"));
     }
+
+    public static Builder builder(final MetricRegistry metricsRegistry) {
+        return new Builder().metricsRegistry(metricsRegistry);
+    }
+
+    public static class Builder {
```
Comment (arteam) [inline metrics-httpclient/src/main/java/com/codahale/metrics/httpclient/InstrumentedHttpClientConnectionManager.java]: Could you please write a unit test for `Builder`? Just with specifying a `MetricsRegistry` and a name as you described in the issue.

**Review by arteam — CHANGES_REQUESTED** [review]
Hi @snakeru, thanks for contributing! I like the idea of adding a builder to `InstrumentedHttpClientConnectionManager`, its constructors indeed are rather cumbersome. I've left some comments regarding the implementation, though.

File: metrics-httpclient/src/main/java/com/codahale/metrics/httpclient/InstrumentedHttpClientConnectionManager.java (hunk 1 above)
Comment (snakeru) [reply metrics-httpclient/src/main/java/com/codahale/metrics/httpclient/InstrumentedHttpClientConnectionManager.java]: ok, will undo

File: metrics-httpclient/src/main/java/com/codahale/metrics/httpclient/InstrumentedHttpClientConnectionManager.java (hunk 2 above)
Comment (snakeru) [reply metrics-httpclient/src/main/java/com/codahale/metrics/httpclient/InstrumentedHttpClientConnectionManager.java]: Arteam, my understanding was that this was just added to the development branch and did not yet made it to any stable release. In this case it is probably ok to make backwards-incompatible changes? Or do you still think that I should add these constructors back?

File: metrics-httpclient/src/main/java/com/codahale/metrics/httpclient/InstrumentedHttpClientConnectionManager.java (hunk 3 above)
Comment (snakeru) [reply metrics-httpclient/src/main/java/com/codahale/metrics/httpclient/InstrumentedHttpClientConnectionManager.java]: ok, will remove the check.

File: metrics-httpclient/src/main/java/com/codahale/metrics/httpclient/InstrumentedHttpClientConnectionManager.java (hunk 4 above)
Comment (snakeru) [reply metrics-httpclient/src/main/java/com/codahale/metrics/httpclient/InstrumentedHttpClientConnectionManager.java]: The whitespacing was changed by the IntelliJ (which, I think, is a good thing, as long as all users have consistently configured autoformatters). But I'll undo it, no problem.

File: metrics-httpclient/src/main/java/com/codahale/metrics/httpclient/InstrumentedHttpClientConnectionManager.java (hunk 5 above)
Comment (snakeru) [reply metrics-httpclient/src/main/java/com/codahale/metrics/httpclient/InstrumentedHttpClientConnectionManager.java]: There are two notes from my side here:
1) I think these improve readability a bit, though I might be mistaken
2) TBH, this is an autogenerated code. I first did it with lombok and then delombokize the code to get the template out.

Again - since this is your repo, not mine, will remove "final", no problem.

File: metrics-httpclient/src/main/java/com/codahale/metrics/httpclient/InstrumentedHttpClients.java (hunk 6 above)
Comment (snakeru) [reply metrics-httpclient/src/main/java/com/codahale/metrics/httpclient/InstrumentedHttpClients.java:31]: If we go for the builder pattern, we should discourage usage of the constructor. Ideally - make it private, which consequently makes this change mandatory. So even if we end up not making the constructor private - this would still be a good change as we showcase, how the library is to be used.

Performance-wise this is not a big deal, as setting up metrics usually happens only once per binary execution.

File: metrics-httpclient/src/main/java/com/codahale/metrics/httpclient/InstrumentedHttpClientConnectionManager.java (hunk 7 above)
Comment (snakeru) [reply metrics-httpclient/src/main/java/com/codahale/metrics/httpclient/InstrumentedHttpClientConnectionManager.java]: Sure.

**Comment by snakeru** [comment]
I pushed out some updates, not all of your comments are addressed (please see my comments above).
Standing by for your feedback.

Also, a quick question - once we settle up on some version - do you want me to squash commits or do you prefer to have the complete history with all the small changes?

File: metrics-httpclient/src/main/java/com/codahale/metrics/httpclient/InstrumentedHttpClientConnectionManager.java (hunk 2 above)
Comment (arteam) [reply metrics-httpclient/src/main/java/com/codahale/metrics/httpclient/InstrumentedHttpClientConnectionManager.java]: Please add them back. Yes, it's a development branch, but we still don't want to break the compatibility across major updates.

**Comment by arteam** [comment]
@snakeru Please add the constructors and it's good to merge. You can organize your commits either way, Github for some time has been providing an option to do a squash and a fast-forward merge automatically.

**Comment by snakeru** [comment]
Ah, @arteam, I thought that this was just added, but now looking at the file history I see that it spans back at least to 2014 and probably more. Adding the constructors back.

Still - should I may be add @Deprecated to them at least?

**Comment by arteam** [comment]
I agree, `@Deprecated` would make sense.

**Comment by snakeru** [comment]
Done

**Comment by arteam** [comment]
Thank you for the contribution!
